### PR TITLE
Mechanik20051988/gh 5419 choose allocator for memtx cpp 14

### PR DESCRIPTION
**Pull request by EvgenyMekhanik**
Add option to choose memtx allocator

### CODE REVIEW

**Review by Korablev77 — COMMENTED** [review]
Why is this test located at root dir? I mean shouldn't it be in test/perf/ ?
Some info concerning the test itself, how to run it, type of workload, reference values would be nice to see.

**Review by Korablev77 — COMMENTED** [review]
Commit says that it changes file extension, but in fact it contains also code changes.
Add to commit message related issue label and explain why it is needed.

**Comment by Korablev77** [comment]
There should be ability to run all memtx tests with both allocators. Otherwise new system allocator is likely to be untested.

**Comment by Korablev77** [comment]
If we used vtabs instead of templates, how severe degradation would be?

File: src/box/memtx_space.cc [hunk 1]
```diff
@@ -320,15 +320,16 @@ dup_replace_mode(uint32_t op)
...
 
+template<class Allocator>
 static int
 memtx_space_execute_replace(struct space *space, struct txn *txn,
 			    struct request *request, struct tuple **result)
 {
 	struct memtx_space *memtx_space = (struct memtx_space *)space;
 	struct txn_stmt *stmt = txn_current_stmt(txn);
 	enum dup_replace_mode mode = dup_replace_mode(request->type);
-	stmt->new_tuple = memtx_tuple_new(space->format, request->tuple,
```
Comment (alyapunov) [inline src/box/memtx_space.cc:330]: Let's try to use `space->format->tuple_new` here.

**Review by alyapunov — CHANGES_REQUESTED** [review]
I don't like the patch.
The original idea was to open a way for implementing different allocators. Now I see that commit, that introduces malloc/free hase hundred lines of code. This burden is too heavy.
It's hard to say exactly what is wrong. Looks like almost everything.
Please try to do it simple and clear.
1. Allocator should have few methods. Alloc/free is a perfect couple. Create/destroy is acceptable.
2. All allocators must have provide their stats in lua. At the same time. Please leave current small statics untouched.
3. I think we should convert functions to templates: memtx_tuple_new, memtx_tuple_delele, memtx_chunk_new, memtx_chunk_delete.
4. Please make performance results readable, measure an put to GH issue result after every significant commit.
5. Please use CAPS for template parameters.

I would try a pattern (just a raw prototype!):
```c++
struct SysAlloc {
    void *alloc() {...}
   ...
};

struct SmallAlloc {
    struct small_alloc alloc;
    void *alloc() {...}
};

struct MemtxEngine {
    struct SysAlloc sysal;
    struct SmallAlloc small;
    template <class ALLOC>
    ALLOC* getAlloc() {
         // That seem to be C++17, but it is also possible to implement in C++9803
         if constexpr(std::is_same_v<ALLOC, SysAlloc>)
             return &sysal;
         else
              return &small;
    }
};

```

File: test/box/moved_delayed_free_mode.test.lua [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    elseif op == 3 then
+        for key = 1, count do
+            space:upsert({key, key + 5000}, {{'=', 2, key + 10000}})
+        end
+    elseif op == 4 then
+        for key = 1, count do
+            space:delete({key})
+        end
+    else
+    	assert(0)
```
Comment (Korablev77) [inline test/box/moved_delayed_free_mode.test.lua:25]: Nit: tab instead of spaces.

File: test/box/moved_delayed_free_mode.test.lua [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+test_run:cmd("setopt delimiter ''");
+
+space = box.schema.space.create('test')
+space:format({ {name = 'id', type = 'unsigned'}, \
+               {name = 'year', type = 'unsigned'} })
+_ = space:create_index('primary', { parts = {'id'} })
+channel = fiber.channel(1)
+
+-- Try to insert/replace/upsert/delete tuples in fiber
+-- in parallel with the snapshot creation.
```
Comment (Korablev77) [inline test/box/moved_delayed_free_mode.test.lua:37]: Actually I don't get the purpose of test. What are you verifying? Now it looks like check that Tarantool doesn't crash. I bet there's already test checking that parallel operations during snapshot are processed correctly.

File: src/box/memtx_engine.c [hunk 4]
```diff
@@ -78,6 +78,57 @@ enum {
 	MAX_TUPLE_SIZE = 1 * 1024 * 1024,
 };
 
+static inline void
+memtx_free_tuple_from_lifo(struct memtx_engine *memtx, void *item)
+{
+	struct memtx_tuple *memtx_tuple = (struct memtx_tuple *)item;
+	struct tuple *tuple = &memtx_tuple->base;
+	size_t total = tuple_size(tuple) +
+		offsetof(struct memtx_tuple, base);
```
Comment (Korablev77) [inline src/box/memtx_engine.c:117]: Nit: offsetof fits into one line with tuple_size

File: src/box/memtx_engine.c [hunk 5]
```diff
@@ -78,6 +78,57 @@ enum {
 	MAX_TUPLE_SIZE = 1 * 1024 * 1024,
 };
 
+static inline void
+memtx_free_tuple_from_lifo(struct memtx_engine *memtx, void *item)
```
Comment (Korablev77) [inline src/box/memtx_engine.c:112]: Function called _from_lifo but in fact it has nothing to do with lifo. I suggest you to rename it to memtx_tuple_free or sort of and use it in memtx_tuple_delete as well.

**Review by Korablev77 — COMMENTED** [review]
We already have some benchmarks imitating production workloads: tpcc, cbench etc.
I'd better check results using them. Anyway 10k operations seem to be small number.

**Review by Korablev77 — COMMENTED** [review]
I guess now you can change commit subject to:
memtx: convert memtx_engine.c to *.cc

Also I wonder why not using .cpp file extension?

File: src/box/memtx_engine.cc [hunk 6]
```diff
@@ -78,55 +81,84 @@ enum {
...
+	return NULL;
+}
+
+template <class ALLOC>
+static inline struct lifo *
+memtx_engine_get_delayed_lifo(struct memtx_engine *memtx)
+{
+	if (std::is_same<ALLOC, SmallAlloc>::value)
+		return &memtx->delayed;
+	return NULL;
```
Comment (Korablev77) [inline src/box/memtx_engine.cc:107]: Before returning null I'd place unreachable()

File: src/box/allocator.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <small/small.h>
+
+struct SmallAlloc
+{
+	static inline void *
```
Comment (Korablev77) [inline src/box/allocator.h:37]: Why don't you want to encapsulate small_alloc into SmallAlloc as a static member? Then you can get rid of memtx_engine_get_alloc() and call directly ALLOC::alloc() without dispatching.
Moreover, I'd add function delayed_free() which would also encapsulate all routines connected with lifo. Otherwise we are publishing delayed mode internals implementation (i.e. any delayed mode implementation has to operate on lifo etc).

**Review by Korablev77 — COMMENTED** [review]
Implement system allocator, based on malloc

Commit subj missing subsystem. I'd add memtx: prefix

File: src/box/sysalloc.c [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	rlist_foreach_entry_safe(item, &alloc->allocations, rlist, tmp)
+		sysfree(alloc, ((void *)item) + sizeof(struct container), item->bytes);
+	assert(alloc->used_bytes == 0);
+}
+
+void *
+sysalloc(struct sys_alloc *alloc, size_t bytes)
+{
+#ifndef _NDEBUG
+	assert(pthread_equal(alloc->thread_id, pthread_self()));
```
Comment (Korablev77) [inline src/box/sysalloc.c:67]: Why do you put assert under ifndef ndebug?

File: src/box/sysalloc.c [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+#endif
+
+	void *ptr = malloc(sizeof(struct container) + bytes);
+	if (!ptr)
```
Comment (Korablev77) [inline src/box/sysalloc.c:71]: ptr != NULL

File: src/box/sysalloc.c [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+void
+sysfree(struct sys_alloc *alloc, void *ptr, size_t bytes)
+{
+#ifndef _NDEBUG
+	assert(alloc->thread_id == pthread_self());
+#endif
+	ptr -= sizeof(struct container);
+	unsigned remaining = bytes % QUOTA_UNIT_SIZE;
```
Comment (Korablev77) [inline src/box/sysalloc.c:106]: bytes of type size_t - let's make remaining be of type size_t as well.

File: src/box/box.cc [hunk 11]
```diff
@@ -1077,6 +1077,20 @@ box_check_sql_cache_size(int size)
 	return 0;
 }
 
+static void
+box_check_allocator(void)
+{
+	const char *allocator = cfg_gets("memtx_allocator");
+	if (allocator == NULL ||
```
Comment (Korablev77) [inline src/box/box.cc:1084]: Default value is "small". Could it really become NULL?

File: src/box/sysalloc.c [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+	 * QUOTA_UNIT_SIZE bytes. Divide the requested bytes into two parts:
+	 * integer number of blocks by QUOTA_UNIT_SIZE bytes and remaining
+	 * bytes count.
+	 */
+	unsigned remaining = bytes % QUOTA_UNIT_SIZE;
+	unsigned units = bytes / QUOTA_UNIT_SIZE;
+	/*
+	 * Check that we need one more quota block for remaining bytes
+	 */
+	if (small_align(alloc->used_bytes, QUOTA_UNIT_SIZE) <
```
Comment (Korablev77) [inline src/box/sysalloc.c:85]: I don't get why do you need these alignments involving unit size, list of allocations etc.
Why not simply use quota_use and quota_release?

File: src/box/sysalloc.c [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+};
+
+void
+sys_alloc_create(struct sys_alloc *alloc, struct quota *quota)
+{
+	alloc->used_bytes = 0;
+	alloc->quota = quota;
+	rlist_create(&alloc->allocations);
+#ifndef _NDEBUG
+	alloc->thread_id = pthread_self();
```
Comment (Korablev77) [inline src/box/sysalloc.c:47]: Hm, is it allocator can be used from several threads?

File: test/perf/tuples-operations.test.lua [hunk 14]
```diff
@@ -11,6 +11,8 @@ env = require('test_run')
 clock = require('clock')
 test_run = env.new()
 
+box.cfg{ wal_mode='none' }
```
Comment (Korablev77) [inline test/perf/tuples-operations.test.lua:14]: Is this change related to the patch?

File: test/box/moved_delayed_free_mode.test.lua (hunk 3 above)
Comment (EvgenyMekhanik) [reply test/box/moved_delayed_free_mode.test.lua:37]: Yes it's check that Tarantool doesn't crash! Other tests check this not with a hundred percent guarantee.

File: test/box/moved_delayed_free_mode.test.lua (hunk 2 above)
Comment (EvgenyMekhanik) [reply test/box/moved_delayed_free_mode.test.lua]: Fixed

File: src/box/memtx_engine.c (hunk 4 above)
Comment (EvgenyMekhanik) [reply src/box/memtx_engine.c]: Fixed

File: src/box/memtx_engine.c (hunk 5 above)
Comment (EvgenyMekhanik) [reply src/box/memtx_engine.c]: Fixed

**Comment by EvgenyMekhanik** [comment]
> We already have some benchmarks imitating production workloads: tpcc, cbench etc.
> I'd better check results using them. Anyway 10k operations seem to be small number.

Fixed, delete this test

**Comment by EvgenyMekhanik** [comment]
> I guess now you can change commit subject to:
> memtx: convert memtx_engine.c to *.cc
> 
> Also I wonder why not using .cpp file extension?

Fixed commit message, in all other places we used *.cc extention instead of *.cpp

File: src/box/memtx_engine.cc (hunk 6 above)
Comment (EvgenyMekhanik) [reply src/box/memtx_engine.cc:113]: Fixed

File: src/box/sysalloc.c (hunk 13 above)
Comment (EvgenyMekhanik) [reply src/box/sysalloc.c:47]: No, small allocator also should be used in one thread. See `slab_cache.h` in `small` subfolder

File: src/box/sysalloc.c (hunk 8 above)
Comment (EvgenyMekhanik) [reply src/box/sysalloc.c:67]: Fixed

File: src/box/sysalloc.c (hunk 9 above)
Comment (EvgenyMekhanik) [reply src/box/sysalloc.c:71]: Fixed

File: src/box/sysalloc.c (hunk 10 above)
Comment (EvgenyMekhanik) [reply src/box/sysalloc.c:106]: Fixed

File: src/box/sysalloc.c (hunk 12 above)
Comment (EvgenyMekhanik) [reply src/box/sysalloc.c:85]: The problem is that quota_use requires takes away the minimum QUOTA_UNIT_SIZE from quota limit. So if you make 5 allocation each for 1 byte only it takes away 5 * QUOTA_UNIT_SIZE bytes from quota limit!

File: src/box/box.cc (hunk 11 above)
Comment (EvgenyMekhanik) [reply src/box/box.cc:1093]: Fixed

File: test/perf/tuples-operations.test.lua (hunk 14 above)
Comment (EvgenyMekhanik) [reply test/perf/tuples-operations.test.lua]: This test is deleted

File: src/box/allocator.h (hunk 7 above)
Comment (EvgenyMekhanik) [reply src/box/allocator.h:37]: I talked with @alyapunov about it. In previous patch version it was so! But as you can see his comment in review
"Allocator should have few methods. Alloc/free is a perfect couple. Create/destroy is acceptable."
If we encapsulate small_alloc into SmallAlloc, we need to implement stats and other methods it. Also for delayed_free. We don't want to dublicate same code for several allocators.

**Comment by Korablev77** [comment]
Superseded by https://github.com/tarantool/tarantool/pull/6121
